### DISCUSSION TITLE
Launching VSCode From Script

### DISCUSSION THREAD

**Discussion by josephcrawfordSRH** [johnlindquist/kit#1378] [Q&A]
Good Morning,

I am working on a script that will list out my projects and allow me to select one, it will then get a list of all branches for the project and allow me to select one.  I then want it to open this branch of the project in VSCode and I have an arg asking if it should open in a new editor or use an existing open one.  However this does not appear to be working.  It keeps opening new editors rather than using the existing editor that is open.  

```javascript
let windowFlag = "-n";

if (await arg("Open in a new window?", ["Yes", "No"]) === "No") {
  console.log('opening in new window');
  windowFlag = "-r";
}

await $`code ${windowFlag} ${projectPath}`;
```

I have debugged this and found that it is in fact using the `-r` when I say `No`, but VSCode keeps opening new editors.

I know this is probably an issue with the VSCode CLI utility and not that of Script Kit but I was wondering is there a better way this could be done in script kit rather than executing a shell command?

Also does Script Kit come with any caching helpers that would create cache files for a script under the cache directory and allow you to check them, etc?

I am going to implement caching for my project list and branches and then another script to refresh the caches so that I am not waiting so long every time the projects/branches are listed.

**Comment by josephcrawfordSRH** [comment]
Apologies it is actually working properly I had a logic flaw.  Is there a better way though?

**Reply by johnlindquist** [reply]
@josephcrawfordSRH 

There's a `db` helper for simple scenarios like this. It can take an async function it will run if the data doesn't exist yet.

> Note: "Actions" are part of v2

```js

let initDb = async (page = 1) => {
  let result = await get(`https://swapi.dev/api/people?page=${page}`)
  return {
    people: result.data.results,
  }
}

let peopleDb = await db(initDb)

let updateDb = async () => {
  let page = 2
  let { people } = await initDb(page)
  peopleDb.people = people
  peopleDb.write()
}

let actions = [
  {
    name: "Refresh DB",
    shortcut: `${cmd}+3`,
    visible: true,
    onAction: async () => {
      await updateDb()
      setChoices(peopleDb.people.map(p => p.name))
    },
  },
]

let name = await arg(
  "Select a person",
  peopleDb.people.map(p => p.name),
  actions
)
```

**Reply by josephcrawfordSRH** [reply]
@johnlindquist oh this is nice, that would keep me from needing a cache file as I currently do.

I created this under `kenvPath('cache/script-name.json)`
```json
{
    "editor": "code",
    "projects": [
        "project_1",
        "project_2",
        "project_3"
    ],
    "branches": {
        "8768896033379111": [ // this key is a hash of the project id as some characters were not working as object id
            "branch_1",
            "branch_2",
            "branch_3"
        ]
    }
}
```
